### PR TITLE
Hook BinaryFormatter replacement in Resx / Clipboard

### PR DESCRIPTION
**Pull request by JeremyKuhne**
This uses the BinaryFormat support classes to handle common framework types before falling back to the BinaryFormatter.

There is some cleanup in the Resx code and simplification in DataObject by adding internal constant strings for known DataFormats.

Also move MultitargetUtil to the normal CriticalException handler.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/9178)

### CODE REVIEW

File: src/System.Windows.Forms/src/System/Windows/Forms/DataObject.BitmapBinder.cs [hunk 1]
```diff
@@ -11,66 +11,52 @@ namespace System.Windows.Forms;
...
-        /// <param name="typeName"></param>
         public override void BindToName(Type serializedType, out string? assemblyName, out string? typeName)
         {
-            // null strings will follow the default codepath in BinaryFormatter
+            // Null values will follow the default codepath in BinaryFormatter.
             assemblyName = null;
             typeName = null;
-            if (serializedType is not null && !serializedType.Equals(typeof(string)) && !serializedType.Equals(typeof(Bitmap)))
+
+            // Bitmap and string types are safe type to serialize/deserialize.
```
Comment (KlausLoeffelmann) [inline src/System.Windows.Forms/src/System/Windows/Forms/DataObject.BitmapBinder.cs:58]:
```suggestion
            // Bitmap and string types are safe types to serialize/deserialize.
```

File: src/System.Windows.Forms/src/System/Windows/Forms/DataObject.OleConverter.cs [hunk 2]
```diff
@@ -353,55 +351,44 @@ private static unsafe Stream ReadByteStreamFromHandle(nint handle, out bool isSe
         }
 
         /// <summary>
-        ///  Creates a new instance of the Object that has been persisted into the
-        ///  handle.
+        ///  Creates a new instance of the Object that has been persisted into the handle.
```
Comment (KlausLoeffelmann) [inline src/System.Windows.Forms/src/System/Windows/Forms/DataObject.OleConverter.cs:354]:
```suggestion
        ///  Creates a new instance of the object that has been persisted into the handle.
```

**Review by KlausLoeffelmann — DISMISSED** [review]
Looks good to me - a few typo nits are what I found.

File: src/System.Windows.Forms/src/System/Resources/ResXDataNode.cs [hunk 3]
```diff
@@ -89,7 +91,7 @@ public ResXDataNode(string name, object? value, Func<Type?, string>? typeNameCon
         {
             throw new InvalidOperationException(string.Format(SR.NotSerializableType, name, valueType.FullName));
         }
-#pragma warning restore SYSLIB0050 // Type or member is obsolete
+#pragma warning restore SYSLIB0050
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Resources/ResXDataNode.cs:94]: I find that comment useful...

File: src/System.Windows.Forms/src/System/Resources/ResXSerializationBinder.cs [hunk 4]
```diff
@@ -35,69 +42,66 @@ internal ResXSerializationBinder(Func<Type?, string>? typeNameConverter)
...
 
-        // Binder couldn't handle it, let the default loader take over.
+        // Hand back what we found or null to let the default loader take over.
         return type;
     }
 
-    // Get the multitarget-aware string representation for the give type.
     public override void BindToName(Type serializedType, out string? assemblyName, out string? typeName)
     {
-        // Normally we don't change typeName when changing the target framework,
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Resources/ResXSerializationBinder.cs:73]: Why did you remove this comment? This is a test case to try to execute this code.

File: src/System.Windows.Forms/src/System/Windows/Forms/DataFormats.cs [hunk 5]
```diff
@@ -16,143 +16,142 @@ namespace System.Windows.Forms;
 /// </summary>
 public static partial class DataFormats
 {
+    internal const string TextConstant = "Text";
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/DataFormats.cs:19]: In Designer we used pattern `internal const string Text = nameof(Text);` in class named for example `DataFormatConstants`

File: src/System.Windows.Forms/src/System/Resources/ResXDataNode.cs [hunk 6]
```diff
@@ -161,12 +163,7 @@ public string Name
         }
         set
         {
-            ArgumentNullException.ThrowIfNull(value, nameof(Name));
-            if (value.Length == 0)
-            {
-                throw new ArgumentException(nameof(Name));
-            }
-
+            ArgumentException.ThrowIfNullOrEmpty(value, nameof(Name));
```
[the remark this answers is not in the document]
Comment (JeremyKuhne) [reply src/System.Windows.Forms/src/System/Resources/ResXDataNode.cs:166]: It does the same thing as above. It throws ArgumentNull if null or Argument if empty.

**Review by Tanya-Solyanik — APPROVED** [review]

File: src/System.Windows.Forms/src/System/Windows/Forms/DataFormats.cs (hunk 5 above)
Comment (KlausLoeffelmann) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataFormats.cs:19]: I, too, deviate sometimes from that pattern when a name (like `Text`) becomes too generic or may collide (even if only by being a "typical class element name") with a class member.

File: src/System.Windows.Forms/src/System/Resources/ResXDataNode.cs (hunk 3 above)
Comment (JeremyKuhne) [reply src/System.Windows.Forms/src/System/Resources/ResXDataNode.cs:94]: I keep it on the matching one above- it's redundant to have the same comment a few lines later.

File: src/System.Windows.Forms/src/System/Windows/Forms/DataFormats.cs (hunk 5 above)
Comment (JeremyKuhne) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataFormats.cs:19]: Collision is a problem here- I wanted discoverability.

File: src/System.Windows.Forms/src/System/Resources/ResXSerializationBinder.cs (hunk 4 above)
Comment (JeremyKuhne) [reply src/System.Windows.Forms/src/System/Resources/ResXSerializationBinder.cs:73]: It makes no sense. Outside of that the code here has nothing explicitly to do with a given scenario- anything like this should live with the type name converter or type resolver implementation. But again, it doesn't make sense.

Serialization always happens with the object instance. In reverse, the deserialization is the data not a desired target type. That is just the way it works, so it is super unclear what this is trying to say. There is no "property" involved here or in the elsewhere in the ResX code directly.
